File: frontend/src/api/papers.ts
/** 试卷管理 API */

import client from './client';
import type { UnifiedResponse, ResponseMeta } from '@/types/common';
import type { Paper, PaperDetail, PaperReviewRequest } from '@/types/paper';

export async function listPapers(page: number = 1, pageSize: number = 20): Promise<{ data: Paper[]; meta: ResponseMeta }> {
  const res = await client.get<UnifiedResponse<Paper[]>>('/papers/', {
    params: { page, page_size: pageSize },
  });
  return { data: res.data.data ?? [], meta: res.data.meta ?? { page, page_size: pageSize, total: 0 } };
}

export async function getPaper(id: number): Promise<Paper> {
  const res = await client.get<UnifiedResponse<Paper>>(`/papers/${id}`);
  if (!res.data.data) throw new Error('获取试卷失败：服务器未返回数据');
  return res.data.data;
}

export async function getPaperDetail(id: number): Promise<PaperDetail> {
  const res = await client.get<UnifiedResponse<PaperDetail>>(`/papers/${id}/detail`);
  if (!res.data.data) throw new Error('获取试卷详情失败：服务器未返回数据');
  return res.data.data;
}

export async function reviewPaper(id: number, data: PaperReviewRequest): Promise<void> {
  await client.put(`/papers/${id}/review`, data);
}

export async function exportPaperLatex(id: number, includeAnswers: boolean = false): Promise<string> {
  const res = await client.post<UnifiedResponse<{ latex: string }>>(
    `/papers/${id}/export`,
    null,
    { params: { format: 'latex', include_answers: includeAnswers } }
  );
  return res.data.data?.latex ?? '';
}

export async function exportPaperPdf(id: number, includeAnswers: boolean = false): Promise<Blob> {
  const res = await client.post(
    `/papers/${id}/export`,
    null,
    {
      params: { format: 'pdf', include_answers: includeAnswers },
      responseType: 'blob',
    }
  );
  return res.data;
}

export async function exportPaperDocx(id: number, includeAnswers: boolean = false): Promise<Blob> {
  const res = await client.post(
    `/papers/${id}/export`,
    null,
    {
      params: { format: 'docx', include_answers: includeAnswers },
      responseType: 'blob',
    }
  );
  return res.data;
}

export async function deletePaper(id: number): Promise<void> {
  await client.delete(`/papers/${id}`);
}

File: backend/app/api/v1/endpoints/papers.py
"""试卷管理端点."""

from fastapi import APIRouter, Depends, Query
from fastapi.responses import Response
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_db
from app.schemas.common import UnifiedResponse
from app.schemas.paper import PaperRead, PaperReviewRequest
from app.services.export_service import ExportService
from app.services.paper_service import PaperService

router = APIRouter()


@router.get("/")
async def list_papers(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    """获取试卷列表."""
    service = PaperService(db)
    papers, total = await service.list_papers(page=page, page_size=page_size)
    return UnifiedResponse(
        data=[PaperRead.model_validate(p) for p in papers],
        meta={"page": page, "page_size": page_size, "total": total},
    )


@router.get("/{paper_id}")
async def get_paper(
    paper_id: int,
    db: AsyncSession = Depends(get_db),
):
    """获取试卷详情."""
    service = PaperService(db)
    paper = await service.get_paper(paper_id)
    if not paper:
        return UnifiedResponse(code=404, message="试卷不存在")
    return UnifiedResponse(data=PaperRead.model_validate(paper))


@router.get("/{paper_id}/detail")
async def get_paper_detail(
    paper_id: int,
    db: AsyncSession = Depends(get_db),
):
    """获取试卷完整详情（含题目内容和分区信息）."""
    service = PaperService(db)
    paper_data = await service.get_paper_detail(paper_id)
    if not paper_data:
        return UnifiedResponse(code=404, message="试卷不存在")
    return UnifiedResponse(data=paper_data)


@router.post("/{paper_id}/export")
async def export_paper(
    paper_id: int,
    format: str = "pdf",
    include_answers: bool = False,
    db: AsyncSession = Depends(get_db),
):
    """导出试卷.

    format: latex | pdf | docx
    include_answers: 是否包含参考答案
    """
    service = PaperService(db)
    paper_data = await service.get_paper_detail(paper_id)
    if not paper_data:
        return UnifiedResponse(code=404, message="试卷不存在")

    export_service = ExportService()

    if format == "latex":
        content = await export_service.export_to_latex(paper_data, include_answers)
        return UnifiedResponse(data={"latex": content})

    if format == "pdf":
        pdf_bytes = await export_service.export_to_pdf_bytes(paper_data, include_answers)
        if pdf_bytes:
            return Response(
                content=pdf_bytes,
                media_type="application/pdf",
                headers={
                    "Content-Disposition": f'attachment; filename="paper_{paper_id}.pdf"'
                },
            )
        return UnifiedResponse(code=500, message="PDF 导出失败（请确认系统已安装 xelatex）")

    if format == "docx":
        docx_bytes = await export_service.export_to_docx(paper_data, include_answers)
        if docx_bytes:
            return Response(
                content=docx_bytes,
                media_type="application/vnd.openxmlformats-officedocument.wordprocessingml.document",
                headers={
                    "Content-Disposition": f'attachment; filename="paper_{paper_id}.docx"'
                },
            )
        return UnifiedResponse(code=500, message="DOCX 导出失败")

    return UnifiedResponse(code=422, message=f"不支持的导出格式: {format}")


@router.put("/{paper_id}/review")
async def review_paper(
    paper_id: int,
    data: PaperReviewRequest,
    db: AsyncSession = Depends(get_db),
):
    """审核试卷."""
    service = PaperService(db)
    paper = await service.get_paper(paper_id)
    if not paper:
        return UnifiedResponse(code=404, message="试卷不存在")
    await service.review_paper(paper_id, data.status)
    return UnifiedResponse(message="审核完成")


@router.delete("/{paper_id}")
async def delete_paper(
    paper_id: int,
    db: AsyncSession = Depends(get_db),
):
    """删除试卷."""
    service = PaperService(db)
    await service.delete_paper(paper_id)
    return UnifiedResponse(message="删除成功")
